router/protocol.go

```diff
@@ -2,58 +2,335 @@ package router
+	// deadlock.  The result channel is of size 1 so that the
+	// goroutine does not linger even if we encounter an error on
+	// the read side.
+	encodeDone := make(chan error, 1)
+	go func() {
+		encodeDone <- enc.Encode(features)
+	}()
+
+	if err := dec.Decode(&res.Features); err != nil {
+		return err
+	}
+
+	if err := <-encodeDone; err != nil {
+		return err
+	}
+
+	if pubKey == nil {
+		if _, present := res.Features["PublicKey"]; present {
+			return ErrExpectedNoCrypto
+		}
+
+		res.setupNoCrypto(enc, dec)
+	} else {
+		remotePubKeyStr, ok := res.Features["PublicKey"]
+		if !ok {
+			return ErrExpectedCrypto
+		}
+
+		remotePubKey, err := hex.DecodeString(remotePubKeyStr)
+		if err != nil {
+			return err
+		}
+
+		res.setupCrypto(params, enc, dec, remotePubKey, privKey)
+	}
+
+	res.Features = filterV1Features(res.Features)
+	return nil
+}
+
+// In the V1 protocol, the intro fields are sent unencrypted.  So we
+// restrict them to an established subset of fields that are assumed
+// to be safe.
+func filterV1Features(intro map[string]string) map[string]string {
+	safe := make(map[string]string)
+	for _, k := range ProtocolV1Features {
+		if val, ok := intro[k]; ok {
+			safe[k] = val
+		}
+	}
+
+	return safe
+}
+
+// The V2 procotol consists of the protocol identification/version
+// header, followed by:
+//
+// - A single "encryption flag" byte: 0 for no encryption, 1 for
+// encryption.
+//
+// - When the connection is encrypted, 32 bytes follow containing the
+// public key.
+//
+// - Then a stream of gobified values.
+//
+// The gobified values are the messages on the connection (encrypted
+// for an encrypted connection).  The first message contains the
+// encoded features map (so in contrast to V1, it will be encrypted on
+// an encrypted connection).
+func (res *ProtocolIntroResults) doIntroV2(params ProtocolIntroParams, pubKey, privKey *[32]byte) error {
+	// Public key exchange
+	var wbuf []byte
+	if pubKey == nil {
+		wbuf = []byte{0}
+	} else {
+		wbuf = make([]byte, 1+len(*pubKey))
+		wbuf[0] = 1
+		copy(wbuf[1:], (*pubKey)[:])
+	}
+
+	// Write in a separate goroutine to avoid the possibility of
+	// deadlock.  The result channel is of size 1 so that the
+	// goroutine does not linger even if we encounter an error on
+	// the read side.
+	writeDone := make(chan error, 1)
+	go func() {
+		_, err := params.Conn.Write(wbuf)
+		writeDone <- err
+	}()
+
+	rbuf := make([]byte, 1)
+	if _, err := io.ReadFull(params.Conn, rbuf); err != nil {
+		return err
+	}
+
+	switch rbuf[0] {
+	case 0:
+		if pubKey != nil {
+			return ErrExpectedCrypto
+		}
+
+		res.setupNoCrypto(gob.NewEncoder(params.Conn),
+			gob.NewDecoder(params.Conn))
+
+	case 1:
+		if pubKey == nil {
+			return ErrExpectedNoCrypto
+		}
+
+		rbuf = make([]byte, len(pubKey))
+		if _, err := io.ReadFull(params.Conn, rbuf); err != nil {
+			return err
+		}
+
+		res.setupCrypto(params, gob.NewEncoder(params.Conn),
+			gob.NewDecoder(params.Conn), rbuf, privKey)
```

Comment:
cosmetic: I'd put the `gob.NewDecoder` on the same line as `gob.NewEncoder`, here and above.